src/output/plugins/OSXOutputPlugin.cxx

```diff
@@ -688,7 +680,44 @@ osx_render(void *vdata,
 {
 	OSXOutput *od = (OSXOutput *) vdata;
 
-	int count = in_number_frames * od->asbd.mBytesPerFrame;
+	if (od->cancel) {
+		od->ring_buffer->reset();
+		od->cancel = false;
+	}
+
+	// If data is not available, fill the buffer with silence,
+	if (!od->ring_buffer->read_available()) {
+#ifdef ENABLE_DSD
+		if (od->dop_enabled) {
+			const auto silence = od->pcm_export->GetSilence();
+			size_t count = in_number_frames * od->asbd.mBytesPerFrame / silence.size;
+			for (size_t i = 0; i < count; ++i) {
+				memcpy((uint8_t *)buffer_list->mBuffers[0].mData + i * silence.size, silence.data, silence.size);
+			}
+			buffer_list->mBuffers[0].mDataByteSize = count * silence.size;
+			return noErr;
+		}
+#endif
+		size_t count = in_number_frames * od->asbd.mBytesPerFrame;
+		bzero(buffer_list->mBuffers[0].mData, count);
+		buffer_list->mBuffers[0].mDataByteSize = count;
+		return noErr;
+	}
+
+	size_t copy_frames =
+		od->ring_buffer->read_available() / od->asbd.mBytesPerFrame;
+	if (copy_frames > in_number_frames) {
+		copy_frames = in_number_frames;
+	}
+#ifdef ENABLE_DSD
+	// In DoP, we always cut off at even frames to preserve the
+	// DoP marker bits pattern.
+	if (od->dop_enabled) {
+		copy_frames = copy_frames / 2 * 2;
```

Comment:
I think instead of this special case, the code should use `PcmExport::GetOutputBlockSize()`, which is a generic way to calculate this.

Reply:
Umm... then how can I get number 2 here? GetOutputBlockSize()/GetFrameSize()?

Reply:
That would be obscure and fragile. Better always use `PcmExport::GetOutputBlockSize()` and never use the frame size.

Reply:
But core audio provide the number of frames in the method signature.